<P> <img src="https://i.ibb.co/gyNf19D/nhslogo.png" alt="nhslogo" border="0" width="100" align="right"><font size="6"><b> CS4132 Data Analytics</b> </font>

# Lab 3.3: Pandas

## (Section 3.5) Operations on Dataframe

Complete the following questions and upload your `.ipynb` file to Coursemology.

Name the file in the following format:

`Lab<Num><YourName>.ipynb`

### Q1

Use the data below to answer the questions.

In [1]:
import pandas as pd
reviews = pd.read_csv("wine.csv", index_col=0) #read in data from wine.csv
reviews.head(5) #display the first few data from the dataframe

country                                        description  \
0     Italy  Aromas include tropical fruit, broom, brimston...   
1  Portugal  This is ripe and fruity, a wine that is smooth...   
2        US  Tart and snappy, the flavors of lime flesh and...   
3        US  Pineapple rind, lemon pith and orange blossom ...   
4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

a) What is the median of the `points` column in the `reviews` DataFrame?

In [3]:
#your solution
reviews.points.median()

86.0

b) What countries are represented in the dataset? (Your answer should not include any duplicates.)

In [4]:
#your solution
reviews.country.unique()

array(['Italy', 'Portugal', 'US', 'Spain', 'France', 'Germany',
       'Argentina', 'Chile', 'Australia', 'Austria'], dtype=object)

c) How often does each country appear in the dataset? Create a Series `reviews_per_country` mapping countries to the count of reviews of wines from that country.

In [6]:
#your solution
reviews_per_country = reviews.groupby('country').country.count()
reviews_per_country

country
Argentina     2
Australia     2
Austria       1
Chile         5
France       14
Germany       4
Italy        24
Portugal      2
Spain         3
US           43
Name: country, dtype: int64

d) Which wine is the "best bargain"? Create a variable `bargain_wine` with the title of the wine with the highest points-to-price ratio in the dataset.

In [82]:
#your solution
bargain_wine = reviews.iloc[(reviews.points / reviews.price).idxmax()].title
bargain_wine

'Henry Fessy 2012 Nouveau  (Beaujolais)'

e) Is a wine more likely to be "tropical" or "fruity"? Create a Series `descriptor_counts` counting how many times each of these two words appears in the description column in the dataset.

In [84]:
#your solution
descriptor_counts = pd.Series([reviews.description.str.count("tropical").sum(), reviews.description.str.count("fruity").sum()], index=['tropical', 'fruity'])
descriptor_counts

tropical    4
fruity      8
dtype: int64

f) We'd like to host these wine reviews on our website, but a rating system ranging from 80 to 100 points is too hard to understand - we'd like to translate them into simple star ratings. A score of 95 or higher counts as 3 stars, a score of at least 85 but less than 95 is 2 stars. Any other score is 1 star.

Also, the Canadian Vintners Association bought a lot of ads on the site, so any wines from Canada should automatically get 3 stars, regardless of points.

Create a series star_ratings with the number of stars corresponding to each review in the dataset.

In [81]:
#your solution
list_star = []
def stars(row) :
    if (row.country == 'Canada'):
        list_star.append('***')
    elif (row.points >= 95):
        list_star.append('***')
    elif(row.points >= 85):
        list_star.append('**')
    else:
        list_star.append('*')

reviews.apply(stars, axis='columns')
star_ratings = pd.Series(list_star)
star_ratings

0     **
1     **
2     **
3     **
4     **
      ..
95    **
96    **
97    **
98    **
99    **
Length: 100, dtype: object

g) For each taster, count how many reviews each person wrote.

In [17]:
#your solution
reviews.groupby('taster_name').taster_name.count()

taster_name
Alexander Peartree     3
Anna Lee C. Iijima     5
Anne Krebiehl MW       1
Jim Gordon             1
Joe Czerwinski         2
Kerin O’Keefe         13
Matt Kettmann          3
Michael Schachner     10
Paul Gregutt           6
Roger Voss            16
Sean P. Sullivan       6
Virginie Boone        16
Name: taster_name, dtype: int64

Hence, find which revewier wrote the most review in this dataset.

In [22]:
#your solution
reviews.groupby('taster_name').taster_name.count().idxmax()

'Roger Voss'

h) What are the minimum and maximum prices for each variety of wine?

In [25]:
#your solution
reviews.groupby('variety').price.agg([min, max])

min    max
variety                                
Aglianico                   32.0   32.0
Albariño                    16.0   20.0
Bordeaux-style Red Blend    46.0   75.0
Bordeaux-style White Blend  15.0   15.0
Cabernet Franc              25.0   25.0
Cabernet Sauvignon          19.0  100.0
Carmenère                   12.0   12.0
Catarratto                  17.0   17.0
Champagne Blend             55.0   58.0
Chardonnay                  12.0   30.0
Chenin Blanc                16.0   16.0
Frappato                    16.0   16.0
Gamay                        9.0   20.0
Gewürztraminer              12.0   30.0
Grillo                      13.0   13.0
Grüner Veltliner            12.0   12.0
Inzolia                     13.0   13.0
Malbec                      13.0   55.0
Meritage                    32.0   55.0
Merlot                       9.0   55.0
Monica                      14.0   14.0
Nerello Mascalese            NaN    NaN
Nero d'Avola                10.0   35.0
Petit Manseng               11.0   11.0
Petit Verdot                22.0   22.0
Petite Sirah                55.0   55.0
Pinot Gris                  14.0   27.0
Pinot Noir                  13.0   69.0
Portuguese Red              15.0   15.0
Primitivo                   11.0   11.0
Red Blend                   12.0   50.0
Riesling                     9.0   24.0
Rosato                      19.0   19.0
Rosé                        20.0   20.0
Sangiovese                  17.0   30.0
Sauvignon Blanc             14.0   26.0
Tempranillo Blend           28.0   28.0
Tempranillo-Merlot          15.0   15.0
Touriga Nacional             NaN    NaN
Vernaccia                   29.0   29.0
Viognier                    22.0   22.0
Viognier-Chardonnay         15.0   15.0
White Blend                 13.0   19.0
Zinfandel                   24.0   24.0

i) What are the most expensive wine varieties? Create a variable `sorted_varieties` containing a copy of the dataframe from the previous question where varieties are sorted in descending order based on minimum price, then on maximum price (to break ties).

In [27]:
#your solution
sorted_varieties = reviews.groupby('variety').price.agg([min, max]).copy().sort_values(by=['min', 'max'], ascending=False)
sorted_varieties

min    max
variety                                
Champagne Blend             55.0   58.0
Petite Sirah                55.0   55.0
Bordeaux-style Red Blend    46.0   75.0
Meritage                    32.0   55.0
Aglianico                   32.0   32.0
Vernaccia                   29.0   29.0
Tempranillo Blend           28.0   28.0
Cabernet Franc              25.0   25.0
Zinfandel                   24.0   24.0
Petit Verdot                22.0   22.0
Viognier                    22.0   22.0
Rosé                        20.0   20.0
Cabernet Sauvignon          19.0  100.0
Rosato                      19.0   19.0
Sangiovese                  17.0   30.0
Catarratto                  17.0   17.0
Albariño                    16.0   20.0
Chenin Blanc                16.0   16.0
Frappato                    16.0   16.0
Bordeaux-style White Blend  15.0   15.0
Portuguese Red              15.0   15.0
Tempranillo-Merlot          15.0   15.0
Viognier-Chardonnay         15.0   15.0
Pinot Gris                  14.0   27.0
Sauvignon Blanc             14.0   26.0
Monica                      14.0   14.0
Pinot Noir                  13.0   69.0
Malbec                      13.0   55.0
White Blend                 13.0   19.0
Grillo                      13.0   13.0
Inzolia                     13.0   13.0
Red Blend                   12.0   50.0
Chardonnay                  12.0   30.0
Gewürztraminer              12.0   30.0
Carmenère                   12.0   12.0
Grüner Veltliner            12.0   12.0
Petit Manseng               11.0   11.0
Primitivo                   11.0   11.0
Nero d'Avola                10.0   35.0
Merlot                       9.0   55.0
Riesling                     9.0   24.0
Gamay                        9.0   20.0
Nerello Mascalese            NaN    NaN
Touriga Nacional             NaN    NaN

j) For each taster, find the average review score given out by that reviewer.

In [28]:
#your solution
reviews.groupby('taster_name').points.mean()

taster_name
Alexander Peartree    87.000000
Anna Lee C. Iijima    86.800000
Anne Krebiehl MW      88.000000
Jim Gordon            86.000000
Joe Czerwinski        86.000000
Kerin O’Keefe         86.923077
Matt Kettmann         86.666667
Michael Schachner     86.200000
Paul Gregutt          86.500000
Roger Voss            86.437500
Sean P. Sullivan      86.333333
Virginie Boone        86.625000
Name: points, dtype: float64

k) Count the number of reviews available for each combination of countries and varieties.

For example, a Gamay produced in France has 5 reviews. Sort the values in descending order based on wine count.

In [37]:
#your solution
reviews.groupby(['country', 'variety'])[['description']].count().sort_values(by='description', ascending=False)

description
country   variety                                
US        Pinot Noir                            6
          Red Blend                             5
          Cabernet Sauvignon                    5
France    Gamay                                 5
Italy     Red Blend                             4
US        Sauvignon Blanc                       4
          Chardonnay                            4
Italy     White Blend                           4
US        Riesling                              3
Italy     Nero d'Avola                          3
US        Bordeaux-style Red Blend              3
Germany   Riesling                              3
Argentina Malbec                                2
US        Meritage                              2
Italy     Sangiovese                            2
France    Gewürztraminer                        2
          Chardonnay                            2
          Champagne Blend                       2
US        Merlot                                2
          Pinot Gris                            2
          Chenin Blanc                          1
          Cabernet Franc                        1
          Petite Sirah                          1
          Albariño                              1
Spain     Tempranillo-Merlot                    1
          Tempranillo Blend                     1
          Albariño                              1
Portugal  Touriga Nacional                      1
US        Viognier                              1
Portugal  Portuguese Red                        1
US        Malbec                                1
Italy     Primitivo                             1
          Vernaccia                             1
France    Pinot Gris                            1
Australia Rosé                                  1
Austria   Grüner Veltliner                      1
Chile     Carmenère                             1
          Merlot                                1
          Petit Verdot                          1
          Pinot Noir                            1
          Viognier-Chardonnay                   1
France    Bordeaux-style White Blend            1
          Petit Manseng                         1
Germany   Gewürztraminer                        1
Italy     Rosato                                1
          Aglianico                             1
          Cabernet Sauvignon                    1
          Catarratto                            1
          Frappato                              1
          Grillo                                1
          Inzolia                               1
          Monica                                1
          Nerello Mascalese                     1
Australia Chardonnay                            1
US        Zinfandel                             1

Hence, which combination of countries and varieties is most common?

In [89]:
#your solution
reviews.groupby(['country', 'variety'])[['description']].count().sort_values(by='description', ascending=False).iloc[0]

description    6
Name: (US, Pinot Noir), dtype: int64

l) Find the average price for each combination of countries and varieties.

In [78]:
#your solution
reviews.groupby(['country', 'variety'])[['price']].mean()

price
country   variety                              
Argentina Malbec                      21.500000
Australia Chardonnay                  18.000000
          Rosé                        20.000000
Austria   Grüner Veltliner            12.000000
Chile     Carmenère                   12.000000
          Merlot                       9.000000
          Petit Verdot                22.000000
          Pinot Noir                  13.000000
          Viognier-Chardonnay         15.000000
France    Bordeaux-style White Blend  15.000000
          Champagne Blend             56.500000
          Chardonnay                  19.500000
          Gamay                       15.250000
          Gewürztraminer              27.000000
          Petit Manseng               11.000000
          Pinot Gris                  27.000000
Germany   Gewürztraminer              12.000000
          Riesling                    14.333333
Italy     Aglianico                   32.000000
          Cabernet Sauvignon          21.000000
          Catarratto                  17.000000
          Frappato                    16.000000
          Grillo                      13.000000
          Inzolia                     13.000000
          Monica                      14.000000
          Nerello Mascalese                 NaN
          Nero d'Avola                19.000000
          Primitivo                   11.000000
          Red Blend                   17.000000
          Rosato                      19.000000
          Sangiovese                  23.500000
          Vernaccia                   29.000000
          White Blend                 16.000000
Portugal  Portuguese Red              15.000000
          Touriga Nacional                  NaN
Spain     Albariño                    16.000000
          Tempranillo Blend           28.000000
          Tempranillo-Merlot          15.000000
US        Albariño                    20.000000
          Bordeaux-style Red Blend    65.333333
          Cabernet Franc              25.000000
          Cabernet Sauvignon          53.600000
          Chardonnay                  17.000000
          Chenin Blanc                16.000000
          Malbec                      55.000000
          Meritage                    43.500000
          Merlot                      38.500000
          Petite Sirah                55.000000
          Pinot Gris                  16.000000
          Pinot Noir                  41.833333
          Red Blend                   29.600000
          Riesling                    15.333333
          Sauvignon Blanc             19.000000
          Viognier                    22.000000
          Zinfandel                   24.000000

Hence for each country, find the variety with the highest average price.

In [95]:
#your solution
reviews.groupby(['country', 'variety'])[['price']].mean().groupby('country').idxmax()

price
country                                  
Argentina             (Argentina, Malbec)
Australia               (Australia, Rosé)
Austria       (Austria, Grüner Veltliner)
Chile               (Chile, Petit Verdot)
France          (France, Champagne Blend)
Germany               (Germany, Riesling)
Italy                  (Italy, Aglianico)
Portugal       (Portugal, Portuguese Red)
Spain          (Spain, Tempranillo Blend)
US         (US, Bordeaux-style Red Blend)

### Q2

Use the data below to answer the questions.

In [52]:
import pandas as pd;
DFAmount = pd.DataFrame(data = [[4,19,13,9],[7,13,20,12],[11,16,13,5],[4,3,4,5]], columns = ['XS','S','M','L'], index = ['Shirt','Tee Shirt','Polo Shirt','V Neck Shirt'])
DFAmount

XS   S   M   L
Shirt          4  19  13   9
Tee Shirt      7  13  20  12
Polo Shirt    11  16  13   5
V Neck Shirt   4   3   4   5

In [71]:
DFSellingprice = pd.DataFrame(data = [[21.9,22.9,23.9,24.9],[23.9,23.9,25.9,25.9],[29,29,32,32],[28,28,28,28]], columns = ['XS','S','M','L'], index = ['Shirt','Tee Shirt','Polo Shirt','V Neck Shirt'])
DFSellingprice

XS     S     M     L
Shirt         21.9  22.9  23.9  24.9
Tee Shirt     23.9  23.9  25.9  25.9
Polo Shirt    29.0  29.0  32.0  32.0
V Neck Shirt  28.0  28.0  28.0  28.0

In [54]:
DFCostprice = pd.DataFrame(data = [[18.0,18,19,19],[20,20,21,21],[25,25,25,25],[20,20,22,22]], columns = ['XS','S','M','L'], index = ['Shirt','Tee Shirt','Polo Shirt','V Neck Shirt'])
DFCostprice

XS   S   M   L
Shirt         18.0  18  19  19
Tee Shirt     20.0  20  21  21
Polo Shirt    25.0  25  25  25
V Neck Shirt  20.0  20  22  22

a) Find the total cost price of the all the item in the shop.

In [60]:
#your solution
(DFAmount * DFCostprice).sum().sum()

3367.0

b) Find the total profit if he sold all the item.

In [61]:
#your solution
(DFAmount * DFSellingprice).sum().sum() - (DFAmount * DFCostprice).sum().sum()

804.2999999999993

c) He realized that he had entered the selling price of the shirt wrongly. The selling price of the shirt in increasing sizes is 22.90, 24.90, 25.90 and 26.90 respectively. Make the necessary amendment to the DFSellingprice.

In [96]:
#your solution
DFSellingprice.loc['Shirt'] = [22.9, 24.9, 25.9, 26.9]
DFSellingprice

XS     S     M     L
Shirt         22.9  24.9  25.9  26.9
Tee Shirt     23.9  23.9  25.9  25.9
Polo Shirt    29.0  29.0  32.0  32.0
V Neck Shirt  28.0  28.0  28.0  28.0